pkg/clustermanager/cluster_manager.go

```diff
@@ -177,7 +233,42 @@ func (c *ClusterManager) DeleteCluster(ctx context.Context, managementCluster, c
 }
 
 func (c *ClusterManager) UpgradeCluster(ctx context.Context, managementCluster, workloadCluster *types.Cluster, clusterSpec *cluster.Spec, provider providers.Provider) error {
-	if err := c.applyCluster(ctx, managementCluster, workloadCluster, clusterSpec, provider, true, types.WithNodeRef(), types.WithNodeHealthy()); err != nil {
+	cpContent, mdContent, err := provider.GenerateClusterApiSpecForUpgrade(ctx, managementCluster, workloadCluster, clusterSpec)
+	if err != nil {
+		return fmt.Errorf("error generating capi spec: %v", err)
+	}
+
+	if err = c.writeCapiSpecFile(clusterSpec.ObjectMeta.Name, templater.AppendYamlResources(cpContent, mdContent)); err != nil {
+		return err
+	}
+
+	err = c.Retrier.Retry(
+		func() error {
+			return c.clusterClient.ApplyKubeSpecFromBytesWithNamespace(ctx, managementCluster, cpContent, constants.EksaSystemNamespace)
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("error applying capi control plane spec: %v", err)
+	}
+
+	if clusterSpec.Spec.ExternalEtcdConfiguration != nil {
+		logger.V(3).Info("Waiting for external etcd to be ready")
+		err = c.clusterClient.WaitForManagedExternalEtcdReady(ctx, managementCluster, etcdWaitStr, workloadCluster.Name)
+		if err != nil {
+			return fmt.Errorf("error waiting for external etcd for workload cluster to be ready: %v", err)
+		}
+		logger.V(3).Info("External etcd is ready")
+		// the condition external etcd ready if true indicates that all etcd machines are ready and the etcd cluster is ready to accept requests
+	}
+
+	logger.V(3).Info("Waiting for control plane to be ready")
+	err = c.clusterClient.WaitForControlPlaneReady(ctx, managementCluster, ctrlPlaneWaitStr, workloadCluster.Name)
+	if err != nil {
+		return fmt.Errorf("error waiting for workload cluster control plane to be ready: %v", err)
+	}
+
+	logger.V(3).Info("Waiting for controlplane and worker machines to be ready")
+	if err = c.waitForNodesReady(ctx, managementCluster, types.WithNodeRef(), types.WithNodeHealthy()); err != nil {
```

Comment:
do you think we should modify this to only check the nodeRef and nodeHealthy for controlplane nodes at this point? and then after applying mdContent run this check for worker nodes/all nodes at that point

Reply:
that makes sense to me. let me test it to see if it does the work